### ISSUE TITLE
Add a subscription example that uses websockets/signalr

### ISSUE THREAD

**Issue by BhaskaranR** [graphql-dotnet/graphql-dotnet#186]

**Comment by joemcbride** [comment]
No, there is not currently an example that uses those.  The `subscription` type has been added fairly recently, so it should now be possible to add one.

**Comment by neooleg** [comment]
Hi @joemcbride, could you please update -- is there an example appeared?

**Comment by joemcbride** [comment]
@neooleg No, I am not aware of an example yet.

**Comment by neooleg** [comment]
Thanks for quick response.

I plan to have deal with subscriptions, so will share my experiments, if it's useful here.

Related question -- is this possible to setup WebSocket connection (e.g. SignalR) and send everything thru it? I mean to use this WS for queries, mutations as well as for subscriptions?

**Comment by joemcbride** [comment]
@neooleg Yes you can use the engine outside of a HTTP endpoint.  The GraphQL engine does not rely on any specific transport, so you can use the `DocumentExecuter` and the `DocumentWriter` with a WebSocket connection.

**Comment by garethrampton** [comment]
@neooleg I'd be really interested in you seeing your experiments here - I've not found any examples of getting subscriptions working  (with .NET).

I've got some ideas how this might work for us via SignalR (and we're already internally using an event driven architecture), but its just at concept stage right now.

**Comment by neooleg** [comment]
@garethrampton as far as I understood -- is this not a simple solution for SignalR and subscriptions in graphql-dotnet working together?

P.S. We're using HTTP until in development, so didn't have a change to set mentioned experiments. Could you please share more info (pitfalls, code or concepts), if possible?

**Comment by OneCyrus** [comment]
i'm interested to get a basic subscription with apollo client working as well. any starting point to get the schema right and how i should push data from the backend?

**Comment by neooleg** [comment]
@OneCyrus,

Personally, I didn't plan to use Apollo in this initiative -- SignalR and graphql-dotnet only.

However, if you will share anything with Apollo it will be interesting. Although, it might be out of scope for this thread, so sharing a link would be the best 😄

**Comment by joemcbride** [comment]
Just a note that the official spec does have info about how subscriptions should work now, so that should be implemented execution engine, which it currently isn't.  http://facebook.github.io/graphql/#sec-Subscription

**Comment by neooleg** [comment]
@joemcbride thanks for sharing.

Is this means that subscriptions will be a part of graphql-dotnet, so they will be used without additional SignalR code?

**Comment by joemcbride** [comment]
@neooleg my first thought is that the underlying transport would be pluggable, so you would still need something like SignalR for it to work.  The core engine would provide the behavior described in the spec though.

**Comment by OneCyrus** [comment]
@joemcbride i was searching this backend implementation in the source. but couldn't find it. is this something you're working on now?

**Comment by joemcbride** [comment]
@OneCyrus No, I am not currently working on it.

**Comment by neooleg** [comment]
@joemcbride sure, it absolutely make sense. My point was a bit different -- will graphql-dotnet do all job for developer on backend regarding pub/sub and further push (when a developer provides e.g. SignalR transport of it)?

Just to be aligned -- do you have an approximate ETA for the subscriptions scope?
Anyway, as soon as I will tackle a solution -- immediately will share this as promised above.

**Comment by lbrolinz** [comment]
@neooleg @joemcbride , Hi, is there any solution for graphql-dotnet subscriptions?

**Comment by joemcbride** [comment]
@lbrolinz This issue will be closed when there is.  It is still Open.

**Comment by neooleg** [comment]
## Disclaimer
Let me to update since I've spent some time for investigation. Probably the summary will be useful for someone, however, it's far from final goal of the topic.

## Abstract
We're targeting .NET Core only, but SignlaR was rewritten from scratch and released just a few days ago. It doesn't have any documentation yet (https://github.com/aspnet/SignalR/issues/724). Hence, IMO it looks like a bit fresh and need to be settled few releases until take it to production.

## Conclusion
Thus, it was decided to postpone implement solution with the graphql-dotnet subscriptions and SignalR 2.0.

**Comment by joemcbride** [comment]
As a small update, #428 by @pekkah has a new `SubscriptionExecutor` which will provide a basis for subscriptions.  Once that is merged, building an example with websockets should be fairly easy. 🎉

**Comment by pekkah** [comment]
I actually have a example ready also. Once I fix those small things with my PR I'll push it out.

**Comment by pekkah** [comment]
Example of using WebSockets with new `ISubscriptionExecuter` can be found [here](https://github.com/graphql-dotnet/subscriptions-transport-ws). Not production ready.

**Comment by diaconesq** [comment]
@joemcbride The link to the spec for subscriptions ( http://facebook.github.io/graphql/#sec-Subscription ) now redirects to http://facebook.github.io/graphql/October2016/ which doesn't seem to contain anything subscription-related. Do you happen to know where the subscription docs moved?

**Comment by junelau** [comment]
It's still on the master spec.

[](https://github.com/facebook/graphql/blob/master/spec/Section%206%20--%20Execution.md)
